### PR TITLE
chore(rest): moving parser api to a rest datasource

### PR DESCRIPTION
**Pull request by bassrock**
## Goal

Refactor parser-graphql-wrapper to use [RestAPIDatasource](https://www.apollographql.com/docs/apollo-server/data/fetching-rest/) so that it is easier to update Pocket's parsing logic within the graph.

This PR aims to streamline how we interact with the Parser so that we can add other datasources into this wrapper.

We are able to remove the itemByUrl dataloader because RestAPIDatasource handles the caching for us and the de-duplicates requests.

### CODE REVIEW

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts [hunk 1]
```diff
@@ -93,32 +84,20 @@ export const resolvers: Resolvers = {
...
-          maxAge: info.cacheControl.cacheHint.maxAge,
-        },
+      const article = await dataSources.parserAPI.getItemData(parent.givenUrl, {
+        videos: MediaTypeParam.AS_COMMENTS,
+        images: MediaTypeParam.AS_COMMENTS,
       });
       // Only extract Marticle data from article string if Parser
       // extracted valid data (isArticle or isVideo is 1)
-      return article.isArticle || article.isVideo
+      return article.isArticle || article.hasVideo == Videoness.HasVideos
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/apollo/resolvers.ts:93]: probably this should check isVide or hasVideos, not just one

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts [hunk 2]
```diff
@@ -167,86 +146,87 @@ export const resolvers: Resolvers = {
...
       } else {
         // Regular URL resolution
-        return getItemByUrl(url);
+        return dataSources.parserAPI.getItemData(url);
       }
     },
     //deprecated
     getItemByItemId: async (_source, { id }, { repositories }) => {
-      return getItemById(id, await repositories.itemResolver);
+      return null; //getItemById(id, await repositories.itemResolver);
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/apollo/resolvers.ts:167]: have the requests stopped for this in studio?

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts [hunk 3]
```diff
@@ -167,86 +146,87 @@ export const resolvers: Resolvers = {
...
       } else {
         // Regular URL resolution
-        return getItemByUrl(url);
+        return await dataSources.parserAPI.getItemData(url);
       }
     },
-    itemByItemId: async (_source, { id }, { repositories }) => {
-      return getItemById(id, await repositories.itemResolver);
+    itemByItemId: async (_source, { id }, { repositories, dataSources }) => {
+      return null; // getItemById(id, await repositories.itemResolver);
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/apollo/resolvers.ts:186]: is this just disabled as part of your draft or intentional?

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts (hunk 2 above)
Comment (bassrock) [reply servers/parser-graphql-wrapper/src/apollo/resolvers.ts:167]: yea they have. gonna remove it

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts (hunk 3 above)
Comment (bassrock) [reply servers/parser-graphql-wrapper/src/apollo/resolvers.ts:186]: part of draft, but given the studio comment removing.

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts (hunk 1 above)
Comment (bassrock) [reply servers/parser-graphql-wrapper/src/apollo/resolvers.ts:93]: So the parserAPI parses into our Item object which doesnt have isVideo. I'll make that take this into account

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts [hunk 4]
```diff
@@ -36,35 +29,10 @@ export const resolvers: Resolvers = {
...
-          const errorData = {
-            itemId: item.itemId,
-            info: info,
-          };
-          serverLogger.error(errorMessage, { error: error, data: errorData });
-          Sentry.addBreadcrumb({ message: errorMessage, data: errorData });
-          Sentry.captureException(error);
-          throw error;
-        }
+        const itemLoaderType = await dataLoaders.itemIdLoader.load(item.itemId);
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/apollo/resolvers.ts:34]: 👍

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts [hunk 5]
```diff
@@ -73,17 +41,19 @@ export const resolvers: Resolvers = {
...
       }
-      const articleText = await (
-        dataSources.parserAPI as ParserAPI
-      ).articleLoader.load({
-        url: parent.givenUrl,
-        options: {
-          imageStyle: MediaTypeParam.DIV_TAG,
-          videoStyle: MediaTypeParam.DIV_TAG,
-          maxAge: info.cacheControl.cacheHint.maxAge,
+      // If the field was requested via refreshArticle we need to clear the cache before we request data
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/apollo/resolvers.ts:44]: 👍

File: servers/parser-graphql-wrapper/src/datasources/ParserAPI.ts [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    output: 'regular',
+  };
+
+  private cache: KeyValueCache<string, KeyValueCacheSetOptions>;
+
+  constructor(datasourceConfig?: DataSourceConfig) {
+    super(datasourceConfig);
+    this.cache = datasourceConfig.cache;
+
+    // super({
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/datasources/ParserAPI.ts:93]: Should this be deleted?

File: servers/parser-graphql-wrapper/src/datasources/ParserAPI.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      url,
+      options,
+    );
+    if (clearCache) {
+      this.clearCache(cacheKey);
+    }
+    const data = await this.get<ParserResponse>(config.parser.dataPath, {
+      params: queryParams,
+      cacheKey: cacheKey,
+      signal: AbortSignal.timeout(config.parser.timeout * 1000),
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/datasources/ParserAPI.ts:127]: yay

File: servers/parser-graphql-wrapper/src/test/utils/parserResponse.ts [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+} from '../../datasources/ParserAPI';
+import { ParserResponse } from '../../datasources/ParserAPITypes';
+import { faker } from '@faker-js/faker';
+import { Videoness } from '../../__generated__/resolvers-types';
+import { merge } from 'lodash';
+import Keyv from 'keyv';
+
+export const fakeArticle: Partial<ParserResponse> = {
+  isArticle: 1,
+  article: `<p>"Not so handsome now", thought Harry as he dipped Hermione in hot sauce. The Death Eaters were dead now, and Harry was hungier than he'd ever been.</p>`,
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/test/utils/parserResponse.ts:18]: it would have been a tragedy to replace this data

Comment (bassrock) [reply servers/parser-graphql-wrapper/src/test/utils/parserResponse.ts:18]: yea it really would have

File: servers/parser-graphql-wrapper/src/datasources/ParserAPI.ts [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        {
+          delayFirstAttempt: false,
+          jitter: 'full',
+          maxDelay: 1000,
+          numOfAttempts: config.parser.retries,
+        },
+      );
+    super({ ...datasourceConfig, fetch: backoffFetch });
+    this.cache = datasourceConfig.cache;
+  }
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/datasources/ParserAPI.ts:108]: @bassrock

**Review by kschelonka — APPROVED** [review]

**Pull request merged**